File: packages/react-native-audio-api/src/core/WaveShaperNode.ts
import AudioNode from './AudioNode';
import type BaseAudioContext from './BaseAudioContext';
import { IWaveShaperNode } from '../jsi-interfaces';
import { WaveShaperOptions } from '../types';
import { validateWaveShaperCurve } from '../utils/validation';

export default class WaveShaperNode extends AudioNode {
  private curveWasSet = false;
  private _curve: Float32Array | null = null;

  constructor(context: BaseAudioContext, options?: WaveShaperOptions) {
    const node = context.context.createWaveShaper(options || {});
    super(context, node, options);
    if (options?.curve) {
      this._curve =
        options.curve instanceof Float32Array
          ? options.curve
          : Float32Array.from(options.curve);
      this.curveWasSet = true;
    }
  }

  get curve(): Float32Array | null {
    return this._curve;
  }

  get oversample(): OverSampleType {
    return (this.node as IWaveShaperNode).oversample;
  }

  set curve(curve: Float32Array | null) {
    validateWaveShaperCurve(curve, this.curveWasSet);

    if (curve !== null) {
      this.curveWasSet = true;
    }

    this._curve = curve;
    (this.node as IWaveShaperNode).setCurve(curve);
  }

  set oversample(value: OverSampleType) {
    (this.node as IWaveShaperNode).oversample = value;
  }
}

File: packages/react-native-audio-api/common/cpp/audioapi/core/effects/WaveShaperNode.cpp
#include <audioapi/core/BaseAudioContext.h>
#include <audioapi/core/effects/WaveShaperNode.h>
#include <audioapi/dsp/VectorMath.h>
#include <audioapi/utils/AudioArray.h>
#include <audioapi/utils/AudioBus.h>
#include <audioapi/HostObjects/utils/NodeOptions.h>

#include <algorithm>
#include <memory>
#include <string>

namespace audioapi {

WaveShaperNode::WaveShaperNode(std::shared_ptr<BaseAudioContext> context, const WaveShaperOptions &options)
    : AudioNode(context, options), oversample_(options.oversample) {

  waveShapers_.reserve(6);
  for (int i = 0; i < channelCount_; i++) {
    waveShapers_.emplace_back(std::make_unique<WaveShaper>(nullptr));
  }
  setCurve(options.curve);
  // to change after graph processing improvement - should be max
  channelCountMode_ = ChannelCountMode::CLAMPED_MAX;
  isInitialized_ = true;
}

OverSampleType WaveShaperNode::getOversample() const {
  return oversample_.load(std::memory_order_acquire);
}

void WaveShaperNode::setOversample(OverSampleType type) {
  std::scoped_lock<std::mutex> lock(mutex_);
  oversample_.store(type, std::memory_order_release);

  for (int i = 0; i < waveShapers_.size(); i++) {
    waveShapers_[i]->setOversample(type);
  }
}

std::shared_ptr<AudioArray> WaveShaperNode::getCurve() const {
  std::scoped_lock<std::mutex> lock(mutex_);
  return curve_;
}

void WaveShaperNode::setCurve(const std::shared_ptr<AudioArray> &curve) {
  std::scoped_lock<std::mutex> lock(mutex_);
  curve_ = curve;

  for (int i = 0; i < waveShapers_.size(); i++) {
    waveShapers_[i]->setCurve(curve);
  }
}

std::shared_ptr<AudioBus> WaveShaperNode::processNode(
    const std::shared_ptr<AudioBus> &processingBus,
    int framesToProcess) {
  if (!isInitialized_) {
    return processingBus;
  }

  std::unique_lock<std::mutex> lock(mutex_, std::try_to_lock);

  if (!lock.owns_lock()) {
    return processingBus;
  }

  if (curve_ == nullptr) {
    return processingBus;
  }

  for (int channel = 0; channel < processingBus->getNumberOfChannels(); channel++) {
    auto channelData = processingBus->getSharedChannel(channel);

    waveShapers_[channel]->process(channelData, framesToProcess);
  }

  return processingBus;
}

} // namespace audioapi
